src/metabase/lib/walk.cljc

```diff
@@ -0,0 +1,112 @@
+(ns metabase.lib.walk
+  "Tools for walking and transforming a query.")
```

Comment:
This new namespace is probably the best new thing in this PR. I've been working on a version of this for a while now, this is one I'm pretty happy with.

These are tools for walking and replacing things in a pMBQL query at certain points of interest, see `walk` below. Lots of our middleware currently rolls its own version of this sort of code, now we have one that can be used everywhere. Using `mbql.u/replace` is slow and buggy since it recurses into stuff like `:source-metadata` or `:lib/metadata` and gets its grubby paws on `:field_refs`; other times we've ran into bugs where we didn't recurse into joins inside previous stages (source queries) etc.

Now those bugs will hopefully be a thing of the past.